src/main/java/net/snowflake/client/core/ArrowSqlInput.java

```diff
@@ -0,0 +1,305 @@
+package net.snowflake.client.core;
+
+import static net.snowflake.client.jdbc.SnowflakeUtil.mapExceptions;
+
+import java.io.InputStream;
+import java.io.Reader;
+import java.math.BigDecimal;
+import java.net.URL;
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.Date;
+import java.sql.NClob;
+import java.sql.Ref;
+import java.sql.RowId;
+import java.sql.SQLData;
+import java.sql.SQLException;
+import java.sql.SQLXML;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Iterator;
+import java.util.List;
+import java.util.TimeZone;
+import net.snowflake.client.core.json.Converters;
+import net.snowflake.client.core.structs.SQLDataCreationHelper;
+import net.snowflake.client.jdbc.FieldMetadata;
+import net.snowflake.client.jdbc.SnowflakeLoggedFeatureNotSupportedException;
+import net.snowflake.client.util.ThrowingBiFunction;
+import org.apache.arrow.vector.util.JsonStringHashMap;
+
+@SnowflakeJdbcInternalApi
+public class ArrowSqlInput implements SFSqlInput {
+
+  private final SFBaseSession session;
+  private final Iterator<Object> structuredTypeFields;
+  private final Converters converters;
+  private final List<FieldMetadata> fields;
+
+  private int currentIndex = 0;
+
+  public ArrowSqlInput(
+      JsonStringHashMap<String, Object> input,
+      SFBaseSession session,
+      Converters converters,
+      List<FieldMetadata> fields) {
+    this.structuredTypeFields = input.values().iterator();
+    this.session = session;
+    this.converters = converters;
+    this.fields = fields;
+  }
+
+  @Override
+  public String readString() throws SQLException {
```

Comment:
IMO there are some differences in specific `read()` methods. I think it's too soon to generify this in any way.

Reply:
After analyzing I'm still not seeing a huge benefit of such refactor. The differences are there, mainly because JsonSqlInput is based on ThrowingTriFunction and ArrowSqlInput on ThrowingBiFunction. I guess we can extract not supported methods to some base class but that's about it.

Reply:
nonetheless, extracted some base sqlinput class.